README.md
<!-- TODO: Remodel to match rewrite properly -->
# Grav Theme Basic

<!-- TODO: Theme screenshot -->

**Basic** is a Grav theme designed to provide a functional, if basic, Grav website that meets accessibility standards and can be extended with other themes and plugins. It is currently required to provide the base for the Leaflet Tour plugin.

<!-- TODO: demo website -->

## Requirements

You will need a Grav site in order to use this theme. The site will need to be running Grav version 1.7.0 or higher.

## Installation

<!-- TODO: Develop skeleton and provide instructions on setting up the skeleton -->

Installing the Basic theme can be done in one of two ways. The [Grav Package Manager (GPM)](https://learn.getgrav.org/cli-console/grav-cli-gpm) installation method enables you to install the theme with the admin panel or a terminal command, while the manual method enables you to do so via a zip file.

<!-- TODO: Add to GPM and remove this statement -->
Note that the GPM method will not be available until the theme has actually been released publicly and added to the GPM.

### GPM Installation (Preferred)

The simplest way to install this theme is via the admin panel, assuming you are using the admin plugin. To install, go to the Themes tab on your dashboard, click the **Add** button, look up this theme, and then click **Install**.

Alternatively, you can install this theme using your system's terminal or command line. From the root of your Grav directory type `bin/gpm install basic`. This will install the Basic theme into your `/user/themes` directory within Grav. Its files will be found under `your-site/grav/user/themes/basic`.

### Manual Installation

To install this theme manually:

<!-- TODO: Link to theme repo and possibly getgrav website -->
1. Download the zip file from the theme repository or by finding the files on the GetGrav website.
2. Upload the file to `your-site/grav/user/themes`.
3. Unzip/extract the file.
4. Rename the folder to `basic`.

The filepath to the theme should now be `your-site/grav/user/themes/basic`.

## Updating

Updates to the Basic theme may be published in the future. As with installation, you can update the theme through the Grav Package Manager (via the admin panel or your system's terminal) or manually.

Please note: Any changes you have made to any of the files in the theme will be overwritten. Any files located elsewhere (for example, a .yaml settings file placed in `user/config/themes`) will remain intact. Therefore, it is strongly discouraged to make any changes directly to theme files.

### GPM Update (Preferred)

The simplest way to update this theme is via the admin panel. To do this, go to the Themes tab on your dashboard and check for updates. The dashboard will indicate if any themes have available updates and will allow you to update them individually or all at once.

Alternatively, you can update this theme using your system's terminal or command line. From the root of your Grav directory type `bin/gpm update basic`. This will check if the Basic theme has any updates. If it does, you will be asked whether or not you wish to update. To continue, type `y` and hit enter.

### Manual Update

To update this theme manually:

1. Delete the `your-site/user/themes/basic` directory.
2. Follow the manual installation directions from this readme.
3. Clear the Grav cache by going to your root Grav directory and entering `bin/grav clear-cache` on the terminal.

Note: If you are using the admin panel, there is also a button to clear the cache in the navigation sidebar.

## Getting Started

If you want to set Basic as the default theme and are using the Admin plugin, you can do so by going to the Themes tab, finding Basic, and clicking **Activate**. Otherwise, set it as the default theme by following these steps:

<!-- TODO: Make it so that the following steps stand out less than the preceding paragraph, since the admin panel is the better option -->

1. Navigate to `your-site/grav/user/config/` in your file system.
2. Open the **system.yaml** file.
3. Change the `theme:` setting to `theme: basic`.
4. Save your changes.
5. Clear the Grav cache by going to your root Grav directory and entering `bin/grav clear-cache` on the terminal.

Once this is done, you should be able to see the new theme on the frontend. Keep in mind any customizations made to a previous theme will not be reflected in the new one.

### Basic Configuration

On the theme configuration page:

<!-- TODO: Remove notes when no longer applicable -->

1. If desired, you can upload a favicon to represent your site.
2. If you want to upload a logo for the site header, refer to the relevant section of this readme. <!-- TODO: link -->
3. You can enable Google Analytics by following the information included in the Features section of this document. (Note: Google Analytics is not yet included)
4. Pick your preferred theme color. (Note: Color customization options are not yet included)
<!-- TODO: Option to set link default behavior -->

Also, make sure to check out this document for general plugin and configuration recommendations, as well as essential accessibility information.

## Features

### Plugins

This theme is designed to work well with the following plugins:

- Markdown Notices
- Highlight
- Breadcrumbs

<!-- TODO: Add additional plugins, possibly. Remove breadcrumbs note. -->

Note: The breadcrumbs plugin integration currently requires a bit more work.

Make sure to turn off any built-in CSS from the plugin configuration pages. The theme makes certain modifications, primarily to improve accessibility, which may be overwritten by the built-in plugin CSS. **Failure to turn off built-in CSS may lead to accessibility violations and/or ugly content.**

The Git Sync plugin is also strongly recommended as a way to backup your work.

### Content Creation

Refer to general Grav instructions on how to add pages and add content to those pages.

You may want a markdown cheatsheet if you have not used it before.

Refer to general accessibility instructions to make sure you create accessible content.

<!-- TODO: Also link to some general info on using grav and some general accessibility info (and to a markdown cheatsheet) -->

### Google Analytics

<!-- TODO: Review instructions to make sure they are up to date -->
<!-- TODO: Add instructions as necessary for ga statement and/or opt out button -->

Requirements:
- Google Analytics account

This theme is set up to work with Google Analytics. If you are not planning to use Google Analytics, you can toggle this option off and ignore the rest of this section.

1. Sign in or create a [Google Analytics](https://analytics.google.com/analytics/web/#/home) account.
2. Select the Admin tab.
3. Select an account from the dropdown menu in the Account column.
4. Click on Create Property in the Property column.
5. Fill out the property settings.
6. Create a new Web Data Stream. This option will be provided immediately after creating a new property, but it can also be selected by clicking on Data Streams under the Property column and choosing Add Stream.
7. From the data stream, copy the Measurement ID (a string beginning with G-)
8. Add the measurement ID to the theme configuration (GA Measurement ID).

### Footer Customization

<!-- TODO: If you are using the theme skeleton, a folder called modules with a page called footer has been provided if you wish to include custom footer text. The footer page currently has no content, but any added content will be displayed in the footer on all pages. Neither the modules folder nor the footer page are routable, so they will not appear in the main navigation or the top/bottom page navigation. -->

<!-- TODO: Link to skeleton -->

<!-- If you do not have the modules folder and footer page, you can set them up manually. -->

1. Add a new page.
    1. Page Title: modules
    2. Parent Page: `<root> /`
    3. Page Template: Default
    4. Visible: No
2. Edit the page configuration.
    1. Under Advanced, set Routable to Disabled.
    2. Can also set Visible to Disabled, although this shouldn't be necessary.
    3. Save the page.
3. Add another new page.
    1. Page Title: footer
    2. Parent Page: `/modules`
    3. Page Template: Default
    4. Visible: No
3. Edit the page configuration.
    1. Under Advanced, set Routable to Disabled.
    2. Can also set Visible to Disabled, although this should not be necessary.
    4. Save the page.

You can add custom footer text or other content using the footer page. Any added conent will be displayed in the footer section of all pages on the website. Note that because neither the modules folder nor the footer page are routable, they will not appear in the site navigation.

### Accessibility

<!-- TODO: Provide notes/documentation for accessibility -->

## Configuration Options

The configuration options for the theme and included page templates are detailed below. The label associated with the option in the admin panel interface will be provided, along with the yaml name in parantheses. Additional notes are included as necessary.

### Theme Configuration

If editing this manually, make sure to copy the `/user/themes/basic/basic.yaml` file to the `/user/config/themes/` folder and modify the copy. If you modify the original theme file directly, any changes will be lost with future updates.

<!-- TODO: Provide table for config options -->

- color scheme
- favicon
- link behavior
- header title (and custom text)
- title url (and custom url)
- logo (and alt text)
- logo url
- google analytics (and id)
- ga statement and opt out

#### Logo Options

There are several options for customizing the site title/logo. By default the site title will be displayed at the top of all pages, and will link to the home page.

1. Keep the default. The site title will be displayed at the top of every page and will function as a link to the home page.
    1. Do not change any of the relevant configuration options.
2. Keep the site title, but make it regular text instead of a link.
    1. Enable the Remove Home URL option.
3. Replace the site title with some other text.
    1. Use the Custom Logo Text option.
    2. You can still choose whether or not the text will link to the site's home page (Remove Home URL).
4. Remove the title text entirely.
    1. Enable the Remove Site Title option.
    2. This will only work if you have added a logo file that will be used instead.

Regardless of how you deal with the logo text, you also have the option to add a logo image. You can either replace any text entirely with the image (using Remove Site Title), or you can display the image in addition to the text.

1. Upload a logo image. Make sure it is a reasonable size.
2. Optionally, add a URL to make the image a link.
3. Add appropriate alt text.

#### Defaults

<!-- TODO: default for ga.statement -->
```yaml
enabled: true
color_scheme: 'default'
favicon:
link_new_tab: false
title:
    option: 'default'
    custom:
    url: 'default'
    custom_url:
logo:
    file:
    alt:
    url:
ga:
    enable: true
    id:
    statement:
    opt_out: true
```

## Additional Information

<!-- TODO: Contribution - SGCI? anyone else? -->

This theme uses the MIT license. Please feel free to modify as you like.

<!-- TODO: link -->
If you find any issues, please report them on the theme repository or by emailing digitalscholarship@ou.edu. If you know how to fix the issue, you are welcome to make a pull request on the repository.

While the theme does not currently make use of them, code has been added to allow for accessible modal dialogs. This code has been copied (and slightly modified) from W3C's provided sample code.


CHANGELOG
# v1.0.0

## mm-dd-yyyy

- first stable release

js/ga.js
// TODO: Mention in documentation in case users want to change this text
const GA_BTN_ALLOW_TEXT = "Allow Google Analytics on This Site";
const GA_BTN_DISABLE_TEXT = "Opt Out of Google Analytics on This Site";
const GA_BTN_DISABLED_STATUS = "You have opted out of Google Analytics.";
const GA_BTN_ALLOWED_STATUS = "Thank you for opting back in to Google Analytics!";

$(document).ready(function() {
    if (typeof ga_property !== "undefined") startGoogleAnalytics();
});

// Code to enable working with Google Analytics (mostly for opting out)
const disable_ga = "ga-disable" + ga_property;

// disable tracking if the opt out cookie already exists
if (document.cookie.indexOf(disable_ga + "=true") > -1) window.sessionStorage.setItem("ga_disabled", true);

function startGoogleAnalytics() {
    // make the opt out/back in button work
    $("#ga-toggle-btn").on("click", function() {
        let allowed = (this.getAttribute("data-tracking") === "false"); // if tracking is disabled then toggling the button enables tracking
        document.cookie = disable_ga + "=true; expires=Thu, 31 Dec " + (allowed ? "2099" : "1998") + "23:59:59 UTC; path=/";
        window.sessionStorage.setItem("ga_disabled", !allowed);
        // modify button and status
        $("#ga-toggle-btn").text(allowed ? GA_BTN_DISABLE_TEXT : GA_BTN_ALLOW_TEXT).attr("data-tracking", allowed);
        $("#ga-status").text(allowed ? GA_BTN_ALLOWED_STATUS : GA_BTN_DISABLED_STATUS);
    });
    // check if google analytics has previously been disabled
    if (window.sessionStorage.getItem("ga_disabled") === "true") {
        $("#ga-toggle-btn").text(GA_BTN_ALLOW_TEXT).attr("data-tracking", "false");
        $("#ga-status").text(GA_BTN_DISABLED_STATUS);
    }
}

js/theme.js
const MOBILE_WIDTH = 599; // anything under 600 pixels wide is counted as a mobile device
const BACK_TO_TOP_Y = 250; // at 250px of scroll, the back to top button will appear

$(document).ready(function() {
    
    // navigation functions
    $("#nav-toggle-btn").on("click", function() {
        this.classList.toggle("expanded");
        $("#" + this.getAttribute("data-toggle")).toggleClass("hide");
        togglePopupButton(this);
    });
    $("#main-nav-list .dd-btn").on("click", function() {
        if (this.hasAttribute("disabled")) return;
        this.parentElement.classList.toggle("expanded");
        togglePopupButton(this);
    });

    // back to top link (make sure it scrolls the user to the top, just in case)
    $("#back-to-top").on("click", function() { window.scrollTo(0, 0); });

    // window scroll function (to prevent overloading) - any extension can modify the doWindowScrolledAction
    let window_scroll_tick = false;
    window.onscroll = function() {
        if (!window_scroll_tick) {
            setTimeout(function() {
                doWindowScrollAction();
                window_scroll_tick = false;
            }, 100);
        }
        window_scroll_tick = true;
    }

    // TODO: Just apply to main, or to header/footer/other?
    // link behavior
    if (link_new_tab) {
        // first change target
        $("main a").add("footer a").attr("target", "_blank");
        // but make sure this doesn't apply to the back to top button
        $("#back-to-top").removeAttr("target");
    }
    modifyBlankLinks($("main").add("footer"));

    // ---------- Supported Plugin Handling ---------- //

    // Markdown Notices (set role for notices)
    $(".notices").attr("role", "note");

    // TODO: Other plugins?
    // Note: Some code for tablists could be found in theme.js for the original basic theme
});

/** 
 * A function for modifying link behavior - any links that open in new tab should identify themselves visually and to AT. Provided here so that it is easier to change and/or add to what it affects. Call after modifying links to have the appropriate target.
 * @param areas - a JQuery object containing all the areas to be searched for this kind of link
 */
function modifyBlankLinks(areas) {
    let links = areas.find("a[target='_blank']");
    links.append("<span class='sr-only'>, Opens in new window</span><i aria-hidden='true' class='fas fa-external-link-alt'></i>");
    // deal with image links
    links.children("img").parent().addClass("img-link-external");

}

/**
 * A function that can be overwritten. Overwriting function can easily manually call toggleBackToTop if it still wants that functionality.
 */
function doWindowScrollAction() {
    toggleBackToTop();
}
/**
 * If the current scroll value is greater than that determined by BACK_TO_TOP_Y, show the button. Note: Checking with both body and documentElement for thoroughness
 */
function toggleBackToTop() {
    if (document.body.scrollTop > BACK_TO_TOP_Y || document.documentElement.scrollTop > BACK_TO_TOP_Y) $("#back-to-top").addClass("active");
    else $("#back-to-top").removeClass("active");
}

/**
 * Convenience function for setting/removing aria-expanded for buttons with aria-haspopup="true". Note that the WAI specifications, as of Jan 19, 2022 recommend not including aria-expanded when the popup content is not shown instead of setting it to false. (Setting to false would still be okay though.)
 * @param btn - The HTML button element
 */
function togglePopupButton(btn) {
    if (btn.hasAttribute("aria-expanded")) btn.removeAttribute("aria-expanded");
    else btn.setAttribute("aria-expanded", "true");
}
/**
 * Simple checker using the viewport width. Returns true for small screens, false for larger.
 */
 function isMobile() {
    return (window.innerWidth < MOBILE_WIDTH);
}

function isClickEvent(e) {
    return (e.type === "click" || e.which === 32 || e.which === 13);
}

js/modal.js
/* This document is copied from WAI's example modal dialog js (https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/js/dialog.js) and utils js (https://w3c.github.io/aria-practices/examples/js/utils.js) . Copyright © 2021 W3C® (MIT, ERCIM, Keio, Beihang). */

/* utils.js */

'use strict';
/**
 * @namespace aria
 */

var aria = aria || {};

/**
 * @description
 *  Key code constants
 */
aria.KeyCode = {
  BACKSPACE: 8,
  TAB: 9,
  RETURN: 13,
  SHIFT: 16,
  ESC: 27,
  SPACE: 32,
  PAGE_UP: 33,
  PAGE_DOWN: 34,
  END: 35,
  HOME: 36,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40,
  DELETE: 46,
};

aria.Utils = aria.Utils || {};

// Polyfill src https://developer.mozilla.org/en-US/docs/Web/API/Element/matches
aria.Utils.matches = function (element, selector) {
  if (!Element.prototype.matches) {
    Element.prototype.matches =
      Element.prototype.matchesSelector ||
      Element.prototype.mozMatchesSelector ||
      Element.prototype.msMatchesSelector ||
      Element.prototype.oMatchesSelector ||
      Element.prototype.webkitMatchesSelector ||
      function (s) {
        var matches = element.parentNode.querySelectorAll(s);
        var i = matches.length;
        while (--i >= 0 && matches.item(i) !== this) {
          // empty
        }
        return i > -1;
      };
  }

  return element.matches(selector);
};

aria.Utils.remove = function (item) {
  if (item.remove && typeof item.remove === 'function') {
    return item.remove();
  }
  if (
    item.parentNode &&
    item.parentNode.removeChild &&
    typeof item.parentNode.removeChild === 'function'
  ) {
    return item.parentNode.removeChild(item);
  }
  return false;
};

aria.Utils.isFocusable = function (element) {
  if (element.tabIndex < 0) {
    return false;
  }

  if (element.disabled) {
    return false;
  }

  switch (element.nodeName) {
    case 'A':
      return !!element.href && element.rel != 'ignore';
    case 'INPUT':
      return element.type != 'hidden';
    case 'BUTTON':
    case 'SELECT':
    case 'TEXTAREA':
      return true;
    default:
      return false;
  }
};

aria.Utils.getAncestorBySelector = function (element, selector) {
  if (!aria.Utils.matches(element, selector + ' ' + element.tagName)) {
    // Element is not inside an element that matches selector
    return null;
  }

  // Move up the DOM tree until a parent matching the selector is found
  var currentNode = element;
  var ancestor = null;
  while (ancestor === null) {
    if (aria.Utils.matches(currentNode.parentNode, selector)) {
      ancestor = currentNode.parentNode;
    } else {
      currentNode = currentNode.parentNode;
    }
  }

  return ancestor;
};

aria.Utils.hasClass = function (element, className) {
  return new RegExp('(\\s|^)' + className + '(\\s|$)').test(element.className);
};

aria.Utils.addClass = function (element, className) {
  if (!aria.Utils.hasClass(element, className)) {
    element.className += ' ' + className;
  }
};

aria.Utils.removeClass = function (element, className) {
  var classRegex = new RegExp('(\\s|^)' + className + '(\\s|$)');
  element.className = element.className.replace(classRegex, ' ').trim();
};

aria.Utils.bindMethods = function (object /* , ...methodNames */) {
  var methodNames = Array.prototype.slice.call(arguments, 1);
  methodNames.forEach(function (method) {
    object[method] = object[method].bind(object);
  });
};

/* dialog.js */

/*
*   This content is licensed according to the W3C Software License at
*   https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document
*/

var aria = aria || {};

aria.Utils = aria.Utils || {};

(function () {
  /*
   * When util functions move focus around, set this true so the focus listener
   * can ignore the events.
   */
  aria.Utils.IgnoreUtilFocusChanges = false;

  aria.Utils.dialogOpenClass = 'has-dialog';

  /**
   * @desc Set focus on descendant nodes until the first focusable element is
   *       found.
   * @param element
   *          DOM node for which to find the first focusable descendant.
   * @returns
   *  true if a focusable element is found and focus is set.
   */
  aria.Utils.focusFirstDescendant = function (element) {
    for (var i = 0; i < element.childNodes.length; i++) {
      var child = element.childNodes[i];
      if (aria.Utils.attemptFocus(child) ||
          aria.Utils.focusFirstDescendant(child)) {
        return true;
      }
    }
    return false;
  }; // end focusFirstDescendant

  /**
   * @desc Find the last descendant node that is focusable.
   * @param element
   *          DOM node for which to find the last focusable descendant.
   * @returns
   *  true if a focusable element is found and focus is set.
   */
  aria.Utils.focusLastDescendant = function (element) {
    for (var i = element.childNodes.length - 1; i >= 0; i--) {
      var child = element.childNodes[i];
      if (aria.Utils.attemptFocus(child) ||
          aria.Utils.focusLastDescendant(child)) {
        return true;
      }
    }
    return false;
  }; // end focusLastDescendant

  /**
   * @desc Set Attempt to set focus on the current node.
   * @param element
   *          The node to attempt to focus on.
   * @returns
   *  true if element is focused.
   */
  aria.Utils.attemptFocus = function (element) {
    if (!aria.Utils.isFocusable(element)) {
      return false;
    }

    aria.Utils.IgnoreUtilFocusChanges = true;
    try {
      element.focus();
    }
    catch (e) {
    }
    aria.Utils.IgnoreUtilFocusChanges = false;
    return (document.activeElement === element);
  }; // end attemptFocus

  /* Modals can open modals. Keep track of them with this array. */
  aria.OpenDialogList = aria.OpenDialogList || new Array(0);

  /**
   * @returns the last opened dialog (the current dialog)
   */
  aria.getCurrentDialog = function () {
    if (aria.OpenDialogList && aria.OpenDialogList.length) {
      return aria.OpenDialogList[aria.OpenDialogList.length - 1];
    }
  };

  aria.closeCurrentDialog = function () {
    var currentDialog = aria.getCurrentDialog();
    if (currentDialog) {
      currentDialog.close();
      return true;
    }

    return false;
  };

  aria.handleEscape = function (event) {
    var key = event.which || event.keyCode;

    if (key === aria.KeyCode.ESC && aria.closeCurrentDialog()) {
      event.stopPropagation();
    }
  };

  document.addEventListener('keyup', aria.handleEscape);

  /**
   * @constructor
   * @desc Dialog object providing modal focus management.
   *
   * Assumptions: The element serving as the dialog container is present in the
   * DOM and hidden. The dialog container has role='dialog'.
   *
   * @param dialogId
   *          The ID of the element serving as the dialog container.
   * @param focusAfterClosed
   *          Either the DOM node or the ID of the DOM node to focus when the
   *          dialog closes.
   * @param focusFirst
   *          Optional parameter containing either the DOM node or the ID of the
   *          DOM node to focus when the dialog opens. If not specified, the
   *          first focusable element in the dialog will receive focus.
   */
  aria.Dialog = function (dialogId, focusAfterClosed, focusFirst) {
    this.dialogNode = document.getElementById(dialogId);
    if (this.dialogNode === null) {
      throw new Error('No element found with id="' + dialogId + '".');
    }

    var validRoles = ['dialog', 'alertdialog'];
    var isDialog = (this.dialogNode.getAttribute('role') || '')
      .trim()
      .split(/\s+/g)
      .some(function (token) {
        return validRoles.some(function (role) {
          return token === role;
        });
      });
    if (!isDialog) {
      throw new Error(
        'Dialog() requires a DOM element with ARIA role of dialog or alertdialog.');
    }

    // Wrap in an individual backdrop element if one doesn't exist
    // Native <dialog> elements use the ::backdrop pseudo-element, which
    // works similarly.
    var backdropClass = 'dialog-backdrop';
    if (this.dialogNode.parentNode.classList.contains(backdropClass)) {
      this.backdropNode = this.dialogNode.parentNode;
    }
    else {
      this.backdropNode = document.createElement('div');
      this.backdropNode.className = backdropClass;
      this.dialogNode.parentNode.insertBefore(this.backdropNode, this.dialogNode);
      this.backdropNode.appendChild(this.dialogNode);
    }
    this.backdropNode.classList.add('active');

    // Disable scroll on the body element
    document.body.classList.add(aria.Utils.dialogOpenClass);

    if (typeof focusAfterClosed === 'string') {
      this.focusAfterClosed = document.getElementById(focusAfterClosed);
    }
    else if (typeof focusAfterClosed === 'object') {
      this.focusAfterClosed = focusAfterClosed;
    }
    else {
      throw new Error(
        'the focusAfterClosed parameter is required for the aria.Dialog constructor.');
    }

    if (typeof focusFirst === 'string') {
      this.focusFirst = document.getElementById(focusFirst);
    }
    else if (typeof focusFirst === 'object') {
      this.focusFirst = focusFirst;
    }
    else {
      this.focusFirst = null;
    }

    // Bracket the dialog node with two invisible, focusable nodes.
    // While this dialog is open, we use these to make sure that focus never
    // leaves the document even if dialogNode is the first or last node.
    var preDiv = document.createElement('div');
    this.preNode = this.dialogNode.parentNode.insertBefore(preDiv,
      this.dialogNode);
    this.preNode.tabIndex = 0;
    var postDiv = document.createElement('div');
    this.postNode = this.dialogNode.parentNode.insertBefore(postDiv,
      this.dialogNode.nextSibling);
    this.postNode.tabIndex = 0;

    // If this modal is opening on top of one that is already open,
    // get rid of the document focus listener of the open dialog.
    if (aria.OpenDialogList.length > 0) {
      aria.getCurrentDialog().removeListeners();
    }

    this.addListeners();
    aria.OpenDialogList.push(this);
    this.clearDialog();
    this.dialogNode.className = 'default_dialog'; // make visible

    if (this.focusFirst) {
      this.focusFirst.focus();
    }
    else {
      aria.Utils.focusFirstDescendant(this.dialogNode);
    }

    this.lastFocus = document.activeElement;
  }; // end Dialog constructor

  aria.Dialog.prototype.clearDialog = function () {
    Array.prototype.map.call(
      this.dialogNode.querySelectorAll('input'),
      function (input) {
        input.value = '';
      }
    );
  };

  /**
   * @desc
   *  Hides the current top dialog,
   *  removes listeners of the top dialog,
   *  restore listeners of a parent dialog if one was open under the one that just closed,
   *  and sets focus on the element specified for focusAfterClosed.
   */
  aria.Dialog.prototype.close = function () {
    aria.OpenDialogList.pop();
    this.removeListeners();
    aria.Utils.remove(this.preNode);
    aria.Utils.remove(this.postNode);
    this.dialogNode.className = 'hidden';
    this.backdropNode.classList.remove('active');
    this.focusAfterClosed.focus();

    // If a dialog was open underneath this one, restore its listeners.
    if (aria.OpenDialogList.length > 0) {
      aria.getCurrentDialog().addListeners();
    }
    else {
      document.body.classList.remove(aria.Utils.dialogOpenClass);
    }
  }; // end close

  /**
   * @desc
   *  Hides the current dialog and replaces it with another.
   *
   * @param newDialogId
   *  ID of the dialog that will replace the currently open top dialog.
   * @param newFocusAfterClosed
   *  Optional ID or DOM node specifying where to place focus when the new dialog closes.
   *  If not specified, focus will be placed on the element specified by the dialog being replaced.
   * @param newFocusFirst
   *  Optional ID or DOM node specifying where to place focus in the new dialog when it opens.
   *  If not specified, the first focusable element will receive focus.
   */
  aria.Dialog.prototype.replace = function (newDialogId, newFocusAfterClosed,
    newFocusFirst) {
    var closedDialog = aria.getCurrentDialog();
    aria.OpenDialogList.pop();
    this.removeListeners();
    aria.Utils.remove(this.preNode);
    aria.Utils.remove(this.postNode);
    this.dialogNode.className = 'hidden';
    this.backdropNode.classList.remove('active');

    var focusAfterClosed = newFocusAfterClosed || this.focusAfterClosed;
    var dialog = new aria.Dialog(newDialogId, focusAfterClosed, newFocusFirst);
  }; // end replace

  aria.Dialog.prototype.addListeners = function () {
    document.addEventListener('focus', this.trapFocus, true);
  }; // end addListeners

  aria.Dialog.prototype.removeListeners = function () {
    document.removeEventListener('focus', this.trapFocus, true);
  }; // end removeListeners

  aria.Dialog.prototype.trapFocus = function (event) {
    if (aria.Utils.IgnoreUtilFocusChanges) {
      return;
    }
    var currentDialog = aria.getCurrentDialog();
    if (currentDialog.dialogNode.contains(event.target)) {
      currentDialog.lastFocus = event.target;
    }
    else {
      aria.Utils.focusFirstDescendant(currentDialog.dialogNode);
      if (currentDialog.lastFocus == document.activeElement) {
        aria.Utils.focusLastDescendant(currentDialog.dialogNode);
      }
      currentDialog.lastFocus = document.activeElement;
    }
  }; // end trapFocus

  window.openDialog = function (dialogId, focusAfterClosed, focusFirst) {
    var dialog = new aria.Dialog(dialogId, focusAfterClosed, focusFirst);
  };

  window.closeDialog = function (closeButton) {
    var topDialog = aria.getCurrentDialog();
    if (topDialog.dialogNode.contains(closeButton)) {
      topDialog.close();
    }
  }; // end closeDialog

  window.replaceDialog = function (newDialogId, newFocusAfterClosed,
    newFocusFirst) {
    var topDialog = aria.getCurrentDialog();
    if (topDialog.dialogNode.contains(document.activeElement)) {
      topDialog.replace(newDialogId, newFocusAfterClosed, newFocusFirst);
    }
  }; // end replaceDialog

}());

// added:
$(document).ready(function() {
  $(".dialog-backdrop").on("click", function(e) {
    closeDialog($(this).children()[1])
  });
  $(".dialog-backdrop").children().on("click", function(e) {
      e.stopPropagation();
  });
})